core/src/poh_recorder.rs

```diff
@@ -90,17 +90,20 @@ impl PohRecorder {
 
     pub fn would_be_leader(&self, within_next_n_ticks: u64) -> bool {
         self.has_bank()
-            || self.start_leader_at_tick.map_or(false, |leader_tick| {
-                let ideal_leader_tick = leader_tick.saturating_sub(self.grace_ticks);
-
-                self.tick_height <= self.last_leader_tick
-                    && self.tick_height >= ideal_leader_tick.saturating_sub(within_next_n_ticks)
-            })
+            || self
+                .leader_first_tick_height
+                .map_or(false, |leader_first_tick_height| {
+                    let ideal_leader_tick_height =
+                        leader_first_tick_height.saturating_sub(self.grace_ticks);
+                    self.tick_height + within_next_n_ticks >= ideal_leader_tick_height
+                        && self.tick_height <= self.leader_last_tick_height
+                })
     }
 
     pub fn leader_after_slots(&self, slots: u64) -> Option<Pubkey> {
+        let current_slot = self.tick_height.saturating_sub(1) / self.ticks_per_slot;
```

Comment:
Instead of `.saturating_sub(1)`,  maybe an explicit `if self.tick_height == 0 { return None; }`?  There should never be a leader at tick = 0 now I think, as that tick doesn't exist.

Reply:
I don't think we want to do that. Even if we are at the first slot (`current_slot == 0`), we should still return what the leader will be after `slots` slots right?

Reply:
kk, yeah.  leader *after* slots.

But wait, it's `leader_after_slot()` not `leader_after_slots()` no?   We're passing in a `slot` not a `slots`.

Reply:
There are two calls to this method:

`banking_stage.rs`
```rust
/// Transaction forwarding
pub const FORWARD_TRANSACTIONS_TO_LEADER_AT_SLOT_OFFSET: u64 = 4;

...

poh.leader_after_slots(FORWARD_TRANSACTIONS_TO_LEADER_AT_SLOT_OFFSET)
```

`poh_recorder.rs`
```rust
pub fn next_slot_leader(&self) -> Option<Pubkey> {
  self.leader_after_slots(1)
}
```

`slots` looks good to me

Reply:
But the name is confusing for sure

Reply:
Going to merge, let me know if you want a follow up for changing the name of the method to something that's more clear  :)

Reply:
Yes please!